### PR TITLE
Add SetMaximumMemoryUsageInMegaByte in AutoMLExperiment

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
We are excited to review your PR.

So we can do the best job, please check:

- [x] There's a descriptive title that will make sense to other developers some time from now. 
- [x] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [x] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [x] You have included any necessary tests in the same PR.

This PR adds `SetMaximumMemoryUsageInMegaByte`, which sets limit to the maximum memory a trial can use, and cancel that running trial if it exceeds the maximum memory limitation.

#6293 
### Example
```csharp
// set maximum memory usage to 8gb
experiment.SetMaximumMemoryUsageInMegaByte(8 * 1024);
```

### How it work
`AutoMLExperiment` monitors the memory and cpu usage periodically (the default period is 2 seconds) using `IPerformanceMonitor`, and once the trial exceed the boundary, `AutoMLExperiment` cancel that trial via `MLContext.CancelExecution` and does clean up.

### Known limitation
This requires estimator support cancel during training by checking if context is still alive from time to time, which is not true for all estimators in ML.Net. 
For example, `LGBM` can't  check if context is alive in unmanaged code part, which makes it can't be cancelled during the time when it's calling into native LightGbm dll. In that situation, tiral memory usage might still exceed maximum limit.

### CODE REVIEW

**Comment by luisquintanilla** [comment]
@LittleLittleCloud looks good to me.

What happens when memory usage hits the specified limit? Is the entire experiment cancelled or only that currently running trial? Does enforcing the memory limit mean that GC kicks in more frequently?

**Comment by LittleLittleCloud** [comment]
> @LittleLittleCloud looks good to me.
> 
> What happens when memory usage hits the specified limit? Is the entire experiment cancelled or only that currently running trial? Does enforcing the memory limit mean that GC kicks in more frequently?

Only current trial will be cancelled

GC will be called every time a trial get completed to clean up and release memory, no matter if that trial is successful or not. So enforcing memory limit won't affect GC kick frequency.

File: src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs [hunk 1]
```diff
@@ -332,7 +338,7 @@ private SweepablePipeline CreateBinaryClassificationPipeline(IDataView trainData
 
     internal class BinaryClassificationRunner : ITrialRunner
```
Comment (tarekgh) [inline src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:339]: >BinaryClassificationRunner [](http://example.com/codeflow?start=19&length=26)

Should we use IDisposable with this class?

File: src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs [hunk 2]
```diff
@@ -346,6 +352,13 @@ public BinaryClassificationRunner(MLContext context, IDatasetManager datasetMana
             _rnd = settings.Seed.HasValue ? new Random(settings.Seed.Value) : new Random();
         }
 
+        public void Dispose()
+        {
+            _context.CancelExecution();
+            _context = null;
+            GC.SuppressFinalize(this);
```
Comment (tarekgh) [inline src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:359]: >SuppressFinalize [](http://example.com/codeflow?start=15&length=16)

I am not seeing this class has a finalizer. what exactly this is for?

Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:359]: Oh this is automatically added by VS when I select "implement IDisposable with disposable pattern", I'll remove that as there's no finalizer for that class.

File: src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs (hunk 1 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:339]: That's a great question here. The IDisposable pattern is mainly because I want to make sure `MLContext.CancelExecuation` get called and set to null after the trial is finished while I don't want to explicitly call it's deconstructor or call GC. But I can go another route if you have any recommendation.

Comment (tarekgh) [reply src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:339]: Dispose pattern is not only for GC finalization or deconstruction. It is used to clean up any object when done using it. You already implemented the Dispose method. Having IDisposable will benefit with the `using ()` code pattern.

File: src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs [hunk 3]
```diff
@@ -421,5 +434,27 @@ public TrialResult Run(TrialSettings settings)
 
             throw new ArgumentException($"The runner metric manager is of type {_metricManager.GetType()} which expected to be of type {typeof(ITrainTestDatasetManager)} or {typeof(ICrossValidateDatasetManager)}");
         }
+
+        public Task<TrialResult> RunAsync(TrialSettings settings, CancellationToken ct)
```
Comment (tarekgh) [inline src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:438]: >RunAsync [](http://example.com/codeflow?start=33&length=8)

I don't know if this is a pattern used in other places but I have some hard time to understand how this is really execute as async? 

`Task.FromResult` will always return a completed task. From the code, `Run(settings)` will always run synchronously and not async. Also, any user calling `await RunAsync(...)`  will always not wait at al as the returned task will be completed. 

am I missing anything here?

Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:438]: `RunAsync` allows `AutoMLExperiment` to pass a cancellation token and cancel running trials if necessary, while `Run` doesn't.

Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:438]: Mostly because of cancellation pattern. `AutoMLExperiment` provides an async api to start training which implements cancellation pattern. During automl training, what `AutoMLExperiment` does is basically kick off trials in a for-loop and check if cancellation token get triggered and that's how cancellation pattern being implemented. So it would be nice if trial runner also supports the cancellation pattern, which accepts a cancellation token and return a Task.

So to summarize,  `ITrialRunner.RunAsync` returning a task is more required by `AutoMLExperiment` class which makes cancellation easier to be done. It's just ML.Net doesn't support async training so we need to wrap it in a task.

Comment (tarekgh) [reply src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:438]: Can we make `ITrialRunner.RunAsync` return non completed task and let the caller await for this task? or can this cause a problem? I mean let `Run(settings)` runs inside a task which we return from `RunAsync` and somehow make this task support cancellation. Or we shouldn't claim `RunAsync` is async method and just rename it to something like `RunWithCancellationToken` and not having it return a task at all.

Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/API/BinaryClassificationExperiment.cs:438]: Oh sure

File: src/Microsoft.ML.AutoML/API/MulticlassClassificationExperiment.cs [hunk 4]
```diff
@@ -424,5 +430,34 @@ public TrialResult Run(TrialSettings settings)
...
+                throw;
+            }
+        }
+
+        public void Dispose()
+        {
+            _context.CancelExecution();
+            _context = null;
+            GC.SuppressFinalize(this);
+        }
```
Comment (tarekgh) [inline src/Microsoft.ML.AutoML/API/MulticlassClassificationExperiment.cs:461]: same previous feedback will apply on RunAsync and Dispose here.

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs [hunk 5]
```diff
@@ -40,31 +41,57 @@ public AutoMLExperiment(MLContext context, AutoMLExperimentSettings settings)
...
         }
 
         private void InitializeServiceCollection()
         {
             _serviceCollection.TryAddTransient((provider) =>
             {
+                var channel = provider.GetRequiredService<IChannel>();
                 var context = new MLContext(_settings.Seed);
-                _cts.Token.Register(() =>
+                _globalCancellationTokenSource.Token.Register(() =>
```
Comment (tarekgh) [inline src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:53]: >_globalCancellationTokenSource [](http://example.com/codeflow?start=16&length=30)

would this instance get created at this point?

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs [hunk 6]
```diff
@@ -40,31 +41,57 @@ public AutoMLExperiment(MLContext context, AutoMLExperimentSettings settings)
...
+
                 return context;
             });
+
+            this.SetPerformanceMonitor(2000);
             _serviceCollection.TryAddSingleton(_settings);
             _serviceCollection.TryAddSingleton(((IChannelProvider)_context).Start(nameof(AutoMLExperiment)));
         }
 
         private void Initialize()
```
Comment (tarekgh) [inline src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:92]: >Initialize [](http://example.com/codeflow?start=21&length=10)

nit: maybe getting rid of this method be better? it has now only one statement.

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs [hunk 7]
```diff
@@ -81,6 +108,12 @@ public AutoMLExperiment SetIsMaximize(bool isMaximize)
             return this;
         }
 
+        public AutoMLExperiment SetMaximumMemoryUsageInMegaByte(double value = double.MaxValue)
```
Comment (tarekgh) [inline src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:111]: >value [](http://example.com/codeflow?start=71&length=5)

what happen if someone passed `NaN` value?

**Comment by tarekgh** [comment]
>         public async Task<TrialResult> RunAsync(CancellationToken ct = default)

what happen if 2 threads called this in same time? or this is not supported scenario?

---
Refers to: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:234 in 16eeeae. [](commit_id = 16eeeae3fa81e3558e431807b809d053e8ac78e8, deletion_comment = False)

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs [hunk 8]
```diff
@@ -187,53 +236,93 @@ public async Task<TrialResult> RunAsync(CancellationToken ct = default)
             ValidateSettings();
             Initialize();
             _settings.CancellationToken = ct;
-            _cts.CancelAfter((int)_settings.MaxExperimentTimeInSeconds * 1000);
-            _settings.CancellationToken.Register(() => _cts.Cancel());
+            _globalCancellationTokenSource.CancelAfter((int)_settings.MaxExperimentTimeInSeconds * 1000);
```
Comment (tarekgh) [inline src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:239]: >)_settings.MaxExperimentTimeInSeconds * 1000 [](http://example.com/codeflow?start=59&length=44)

Any concern with overflowing this calculation?

**Comment by tarekgh** [comment]
>         public async Task<TrialResult> RunAsync(CancellationToken ct = default)

Also, Did we think making _globalCancellationTokenSource not class field and only having it as local inside RunAsync? The current code cause problems in multi-threading scenarios.

---
In reply to: [1233449717](https://github.com/dotnet/machinelearning/pull/6305#issuecomment-1233449717) [](http://example.com/codeflow?ancestors=1233449717)

---
Refers to: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:234 in 16eeeae. [](commit_id = 16eeeae3fa81e3558e431807b809d053e8ac78e8, deletion_comment = False)

File: src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Text;
+using System.Threading.Tasks;
+using System.Timers;
+using Microsoft.ML.Runtime;
+
+namespace Microsoft.ML.AutoML
+{
+    public interface IPerformanceMonitor : IDisposable
```
Comment (tarekgh) [inline src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:15]: >IPerformanceMonitor [](http://example.com/codeflow?start=21&length=19)

Do we expect extending this interface later? I am asking to know if it is better to make it abstract class if we run on .NET Framework and avoid breaking future change.

Comment (tarekgh) [inline src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:15]: Could you please add some comments to the exposed APIs?

File: src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+        private void SampleCpuAndMemoryUsage()
+        {
+            // calculate CPU usage in %
+            using (var process = Process.GetCurrentProcess())
+            {
+                var currentCpuProcessorTime = Process.GetCurrentProcess().TotalProcessorTime;
+                var elapseCpuProcessorTime = currentCpuProcessorTime - _totalCpuProcessorTime;
+                var cpuUsedMs = elapseCpuProcessorTime.TotalMilliseconds;
+                var cpuUsageInTotal = cpuUsedMs / (Environment.ProcessorCount * _checkIntervalInMilliseconds);
```
Comment (tarekgh) [inline src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:101]: >_checkIntervalInMilliseconds [](http://example.com/codeflow?start=80&length=28)

why multiply by _checkIntervalInMilliseconds?

Comment (tarekgh) [inline src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:101]: >Environment.ProcessorCount * _checkIntervalInMilliseconds [](http://example.com/codeflow?start=51&length=57)

nit: This can be calculated once in the constructor.

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs [hunk 11]
```diff
@@ -187,53 +236,93 @@ public async Task<TrialResult> RunAsync(CancellationToken ct = default)
...
+                    {
+                        performanceMonitor.MemoryUsageInMegaByte += (o, m) =>
+                        {
+                            if (_settings.MaximumMemoryUsageInMegaByte is double d && m > d && !trialCancellationTokenSource.IsCancellationRequested)
+                            {
+                                logger.Trace($"cancel current trial {setting.TrialId} because it uses {m} mb memory and the maximum memory usage is {d}");
+                                trialCancellationTokenSource.Cancel();
+
+                                GC.AddMemoryPressure(Convert.ToInt64(m) * 1024 * 1024);
+                                GC.Collect();
```
Comment (tarekgh) [inline src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:280]: why we are doing that?

File: src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs (hunk 10 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:101]: the % of cpu is calcuated by $(CPU time used by current process) / $(total CPU time)

for every `_checkIntervalInMilliseconds`
process uses `TotalProcessorTime - (last total processor time)`
total CPU time is # of processor count * `_checkIntervalInMilliseconds`

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs (hunk 11 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:280]: You mean `GC.Collect` or `GC.AddMemoryPressure` part?

Comment (tarekgh) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:280]: yes

File: src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs (hunk 10 above)
Comment (tarekgh) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:101]: ok. thanks!

Comment (tarekgh) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:101]: comment will be good here.

File: src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs (hunk 9 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:15]: Actually maybe we should make it internal? I don't expect user to provide their own implementation of that interface.

Comment (tarekgh) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:15]: agree.

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs (hunk 6 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:92]: ### Resolved

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs (hunk 8 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:239]: Thanks for catching that, switch to `TimeSpan.FromSeconds` to avoid overflow in int

Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:239]: ### Resolved

File: src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs (hunk 10 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:101]: ### Resolved

Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:101]: compiler should be able to optimize that.

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs (hunk 11 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:280]: This is because if we hit that code, it means the current trial uses memory that exceed the limit. Therefore we would like to run a GC to collect memory after that trial get cancelled.

`GC.AddMemoryPressure` is mostly for LightGbm which might allocate large chunk of unmanaged memory.

Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:280]: ### Resolved

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs (hunk 7 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:111]: I will add a check here to make sure it's either not `NaN` or non-positive.

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs (hunk 11 above)
Comment (tarekgh) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:280]: Wouldn't GC kick off automatically when it needs more memory? at that time will collect any stale objects. I am asking when you are forcing that to happen at this time?

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs (hunk 5 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:53]: Yep

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs (hunk 11 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:280]: But it wouldn't if GC thinks it still has enough memory. And in that case, if we didn't run GC, the next trial will likely to be canceled even if it's not using a lot of memory because the memory used by the last trial is not released.

Comment (tarekgh) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:277]: > But it wouldn't if GC thinks it still has enough memory.

Could you please talk more about that? Why will GC think there will not be enough memory? I expect GC will run to collect any unneeded objects at that time which should free a lot of memory. no? I am not objecting here, I am trying to understand the case.

Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:277]: > I expect GC will run to collect any unneeded objects at that time which should free a lot of memory. no?

Only if GC thinks there's not enough memory and runs a collection, but that's not always the case when a trial gets canceled. For example, if the user sets a very small number for maximum memory usage.

File: src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs (hunk 10 above)
Comment (tarekgh) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:106]: Are you assuming that? or you have tried it? 

Trying quickly in [csharplab](https://sharplab.io/#v2:C4LghgzgtgPgAgJgAwFgBQiCM71wMwAEiBAKgKYTAHoDe6BDBADgE4CWAbmMGQWwHZUA+gDMA3PUatO3XgGMA9v0p9BBAHIBXKAQC8BZBLSNqxxvlIVgACgCUkhnTMmGovQUxIjJgL4PTJhZwACwEAArsggBqYAA2mmR2/k4ujAD0aapUHO5uAFQEAKL8HGwsSlBkggB0EQpyFBAKLADCCpqC3qlZBDn6+RraXalwmACc1hy2wwR+zgHmCAjJ/ib8ZADulpR2tZHAMfGJ0/5zcyb+QA=) I am seeing it is not optimized. note I didn't use `Environment.ProcessorCount` because it is just 1 and will be optimized at that time. so, I used other constant. 

Anyway, it is a minor point, and you can ignore it if you are ok with the current code.

File: src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs (hunk 11 above)
Comment (tarekgh) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/AutoMLExperiment.cs:277]: I admit I am not fully understanding the whole picture :-) I'll leave it to you if you are sure about that. You may ignore my comment then.

File: src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs (hunk 10 above)
Comment (LittleLittleCloud) [reply src/Microsoft.ML.AutoML/AutoMLExperiment/IPerformanceMonitor.cs:106]: Yeah indeed it's not, turns out I'm too optimistic on the capability of how much optimization compiler can be.

Thanks for introducing csharplab, that's such a wonderful tool

**Review by tarekgh — APPROVED** [review]
